File: Julia/gbm.jl
using EvoTrees, CSV, DataFrames, CategoricalArrays, Statistics
using EvoTrees: importance

# Load data
data = DataFrame(CSV.read("C:/Users/Hyrum Hansen/Documents/github/model-benchmarking/data/classification_data.csv", DataFrame))
data[!, :y] = data[!, :y]

y_train, x_train = data[!, :y], Matrix(select(data, Not([:Column1, :y])))

params = EvoTreeRegressor(
    loss=:linear, metric=:mse,
    nrounds=100, eta=0.1,
    max_depth = 3, min_weight = 1.0,
    rowsample=0.8, colsample=0.8)

# Initialize an empty vector to store the fit times
fit_times = Float64[]

# Loop over n iterations and fit the decision tree model
for i in 1:500

    t0 = time()
    model = fit_evotree(params; x_train, y_train)
    t1 = time()

    fit_time = t1 - t0
    push!(fit_times, fit_time)
    
end

# Calculate the total time, median time, minimum time, maximum time, and average time
total_time = sum(fit_times)
median_time = median(fit_times)
min_time = minimum(fit_times)
max_time = maximum(fit_times)
avg_time = mean(fit_times)

# Print the results to the console
println("Total time: ", total_time)
println("Median time: ", median_time)
println("Minimum time: ", min_time)
println("Maximum time: ", max_time)
println("Average time: ", avg_time)

File: R/gbm.R
library(xgboost)

# load the example dataset "data" which has 4 predictors and 3 classes
data <- read.csv("data/classification_data.csv")[,c(2:5)]

#Split the data into features and target
features <- data[,c("x1", "x2", "x3")]
target <- data$y

# Convert the data to a DMatrix object
dtrain <- xgb.DMatrix(data = as.matrix(features), label = target)

# Set the hyperparameters for the gradient boosting model
params <- list(objective = "reg:squarederror",
               booster = "gbtree",
               eta = 0.1,
               max_depth = 3,
               subsample = 0.8,
               colsample_bytree = 0.8)

# Train the gradient boosting model with 100 iterations
model <- xgb.train(params = params, data = dtrain, nrounds = 100)

# Initialize an empty vector to store the fit times
fit_times <- numeric()

# Loop over 500 iterations and fit the decision tree model
for (i in 1:500) {
  # Record the start time
  t0 <- Sys.time()

  # Fit the decision tree model
  model <- xgb.train(params = params, data = dtrain, nrounds = 100)

  # Record the end time
  t1 <- Sys.time()

  # Calculate the fit time and append it to the fit_times vector
  fit_time <- as.numeric(difftime(t1, t0, units = "secs"))
  fit_times <- c(fit_times, fit_time)
}

# Calculate the total time, median time, minimum time, maximum time, and average time
total_time <- sum(fit_times)
median_time <- median(fit_times)
min_time <- min(fit_times)
max_time <- max(fit_times)
avg_time <- mean(fit_times)

# Print the results to the console
cat("Total time: ", total_time, "\n")
cat("Median time: ", median_time, "\n")
cat("Minimum time: ", min_time, "\n")
cat("Maximum time: ", max_time, "\n")
cat("Average time: ", avg_time, "\n")

